pkg/sql/set_cluster_setting.go

```diff
@@ -775,3 +792,54 @@ func toSettingString(
 		return "", errors.Errorf("unsupported setting type %T", setting)
 	}
 }
+
+// unsafeSettingInterlockInfo contains information about the current
+// session that is used by the unsafe setting interlock system.
+type unsafeSettingInterlockInfo struct {
+	sessionID    clusterunique.ID
+	interlockKey string
+}
+
+func (p *planner) makeUnsafeSettingInterlockInfo() unsafeSettingInterlockInfo {
+	return unsafeSettingInterlockInfo{
+		sessionID:    p.ExtendedEvalContext().SessionID,
+		interlockKey: p.SessionData().UnsafeSettingInterlockKey,
+	}
+}
+
+const interlockKeySessionVarName = "unsafe_setting_interlock_key"
+
+// unsafeSettingInterlock ensures that changes to unsafe settings are
+// doubly confirmed by the operator by a special value in a session
+// variable.
+func unsafeSettingInterlock(
+	ctx context.Context,
+	st *cluster.Settings,
+	setting settings.Setting,
+	encodedValue string,
+	info unsafeSettingInterlockInfo,
+) error {
+	// The interlock key is a combination of:
+	// - the session ID, so that different sessions need different keys.
+	// - the setting key, so that different settings need different
+	//   interlock keys.
+	// - the value.
+	h := fnv.New32()
+	h.Write([]byte(info.sessionID.String()))
+	h.Write([]byte(setting.InternalKey()))
+	h.Write([]byte(encodedValue))
```

Comment:
personally I'd remove the value too, so once you've been warned and say you're going to mess with setting X, we don't keep yelling at you to get a new key if you want to set it to 5, 10, then 20.